handler/oauth2/strategy_jwt.go

```diff
@@ -36,66 +37,98 @@ func (h RS256JWTStrategy) AuthorizeCodeSignature(token string) string {
 	return h.signature(token)
 }
 
+func (h *RS256JWTStrategy) ValidateJWT(tokenType fosite.TokenType, token string) (requester fosite.Requester, err error) {
+	t, err := h.validate(token)
+	if err != nil {
+		return nil, err
+	}
+
+	claims := jwt.JWTClaims{}
+	claims.FromMapClaims(t.Claims.(jwtx.MapClaims))
+
+	requester = &fosite.Request{
+		Client: &fosite.DefaultClient{
+			ID: claims.Audience,
+		},
+		Session: &JWTSession{
+			JWTClaims: &claims,
+			JWTHeader: &jwt.Headers{
+				Extra: make(map[string]interface{}),
+			},
+			ExpiresAt: map[fosite.TokenType]time.Time{
+				tokenType: claims.ExpiresAt,
+			},
+			Subject: claims.Subject,
+		},
+	}
+
+	return
+}
+
 func (h *RS256JWTStrategy) GenerateAccessToken(_ context.Context, requester fosite.Requester) (token string, signature string, err error) {
 	return h.generate(fosite.AccessToken, requester)
 }
 
 func (h *RS256JWTStrategy) ValidateAccessToken(_ context.Context, _ fosite.Requester, token string) error {
-	return h.validate(token)
+	_, err := h.validate(token)
+	return err
 }
 
 func (h *RS256JWTStrategy) GenerateRefreshToken(_ context.Context, requester fosite.Requester) (token string, signature string, err error) {
 	return h.generate(fosite.RefreshToken, requester)
 }
 
 func (h *RS256JWTStrategy) ValidateRefreshToken(_ context.Context, _ fosite.Requester, token string) error {
-	return h.validate(token)
+	_, err := h.validate(token)
+	return err
 }
 
 func (h *RS256JWTStrategy) GenerateAuthorizeCode(_ context.Context, requester fosite.Requester) (token string, signature string, err error) {
 	return h.generate(fosite.AuthorizeCode, requester)
 }
 
 func (h *RS256JWTStrategy) ValidateAuthorizeCode(_ context.Context, requester fosite.Requester, token string) error {
-	return h.validate(token)
+	_, err := h.validate(token)
+	return err
 }
 
-func (h *RS256JWTStrategy) validate(token string) error {
-	t, err := h.RS256JWTStrategy.Decode(token)
-	if err != nil {
-		return err
+func (h *RS256JWTStrategy) validate(token string) (t *jwtx.Token, err error) {
+	t, err = h.RS256JWTStrategy.Decode(token)
+
+	if err == nil {
+		err = t.Claims.Valid()
 	}
 
-	// validate the token
-	if err = t.Claims.Valid(); err != nil {
+	if err != nil {
 		if e, ok := err.(*jwtx.ValidationError); ok {
```

Comment:
This entire block is dead code btw. I didn't change it because I was going for minimal impact and didn't want to change the behavior, but the `RS256JWTStrategy.Decode` method already does all of the validation, and it _always_ returns a new string error.

![screen shot 2017-01-21 at 4 45 57 am](https://cloud.githubusercontent.com/assets/1731074/22174502/28a4964e-df95-11e6-8f2c-b9d75a2f328c.png)